### DISCUSSION TITLE
PowerOcean support 1.1.0-pre6

### DISCUSSION THREAD

**Discussion by niltrip** [tolwi/hassio-ecoflow-cloud#301] [Q&A]
Hello, I have now integrated the PowerOcean via this integration. For now, I have 27 parameters. I don't know if I did it right. From my point of view, the structure of the data via the official API is more convoluted than with other products (I did not have other products). I had to extend the method _to_plain. I couldn't find another place. 

For this reason, when initializing the integration, only individual values are displayed (not nested). The others are only updated after a message has been sent by the MQTT broker. As far as I understood? I did not find any suiteable methode like _to_plain at the startup of the integration?

Unfortunately, certain values (mpptPwr, bpPwr, sysLoadPwr, sysGridPwr) are only provided in the quota/all query during initialization. After that, the values are not updated except for bpSoc. Can someone please take a look at this who also has a PowerOcean? It works for me. Or could someone look at my code to see if I've done it in the right way? I was able to identify the following payloads:

```
quota_all {'pcsBPhase': {'vol': 234.69943, 'amp': 0.4840443, 'actPwr': -69.39399, 'reactPwr': 89.9475, 'apparentPwr': 113.60493}, 'bpSoc': 72, 'mpptHeartBeat': [{'mpptPv': [{'vol': 21.821943, 'amp': 0.0, 'pwr': 0.0}, {'vol': 15.705534, 'amp': 0.0, 'pwr': 0.0}]}], 'mpptPwr': 0.0, 'bpPwr': -458.0, 'pcsCPhase': {'vol': 236.6824, 'amp': 0.6625455, 'actPwr': -27.3891, 'reactPwr': 154.40245, 'apparentPwr': 156.81287}, 'emsErrCode': {'errCode': []}, 'sysLoadPwr': 454.0, 'sysGridPwr': -4.0, 'pcsAPhase': {'vol': 229.83286, 'amp': 1.0821258, 'actPwr': 124.57848, 'reactPwr': 215.25775, 'apparentPwr': 248.70805}}



{"cmdId":1,"cmdFunc":96,"param":{"pcsBPhase":{"vol":240.28699,"amp":5.061869,"actPwr":-1212.2328,"reactPwr":99.40008,"apparentPwr":1216.3013},"mpptHeartBeat":[{"mpptPv":[{"vol":345.45685,"amp":8.0504265,"pwr":2781.075},{"vol":239.86316,"amp":4.653842,"pwr":1116.2853}]}],"pcsCPhase":{"vol":240.54668,"amp":4.966705,"actPwr":-1187.8181,"reactPwr":128.27461,"apparentPwr":1194.7244},"pcsAPhase":{"vol":237.46803,"amp":3.9315531,"actPwr":-903.5458,"reactPwr":235.04866,"apparentPwr":933.61816}},"addr":"ems"}

# a lot messages but empty. Should be mpptPwr, bpPwr, sysLoadPwr, sysGridPwr (only in quota/all > old values, look like depends on open app)
{"cmdId":8,"cmdFunc":96,"param":{},"addr":"ems_change_report"}

# bpSoc by change depends on batterie (like in my integration, bpSoc within JTS1_EMS_CHANGE_REPORT)
{"cmdId":8,"cmdFunc":96,"param":{"bpSoc":100},"addr":"ems_change_report"}

# empty
{"cmdId":13,"cmdFunc":96,"param":{},"addr":"ems_param_change_report"}
{"cmdId":11,"cmdFunc":96,"param":{},"addr":"ems_eco_logy_dev"}
{"cmdId":3,"cmdFunc":96,"param":{"emsErrCode":{"errCode":[]}},"addr":"error_code"}
{"cmdId":1,"cmdFunc":224,"param":{},"addr":"ems_logy_dev"}
# looks like batteries but empty
{"cmdId":7,"cmdFunc":96,"param":{},"addr":"bp_addr"}
```

![image](https://github.com/user-attachments/assets/b0336ace-d7b7-4cc1-82a0-5f57443113a8)

**Comment by nielshh** [comment]
Hi Niltrip, I have a powerocean and I'm willing to test it out, But can you give me some pointers on what/how you installed this?

Regards Niels

**Comment by niltrip** [comment]
Hi @nielshh, you have to install my repo. It is a fork of tolwi.
https://github.com/niltrip/hassio-ecoflow-cloud

You can also try another integration for daily use:
https://github.com/niltrip/powerocean

**Comment by nielshh** [comment]
Hi niltrip,

I tried, but when I choose the application API it only shows me a list of devices which doesn't include the powerocean:
<img width="284" alt="Screenshot 2024-10-22 at 09 51 58" src="https://github.com/user-attachments/assets/7b2250f0-6ace-4c62-bbc2-2882c48efa58">

The second integration you mentioned only shows me my SN and firmware version, no other data:
<img width="1218" alt="Screenshot 2024-10-22 at 09 53 36" src="https://github.com/user-attachments/assets/f9774d00-5303-4af2-9b34-93ea164fca87">

Any pointers?

BR Niels

**Comment by niltrip** [comment]
Hi, 

for the first integration you have to use the public API not the application API.

For the second integration i have to idea at the moment. Do you know the Webportal for your PowerOcean (https://portal.ecoflow.com/user/eu/en/login)? Do you use it, or only the App?

Best regards 
Christian

**Reply by nielshh** [reply]
> Hi,
> 
> for the first integration you have to use the public API not the application API.
> 

I will register for the public API and give it a try.

> For the second integration i have to idea at the moment. Do you know the Webportal for your PowerOcean (https://portal.ecoflow.com/user/eu/en/login)? Do you use it, or only the App?
> 
> Best regards Christian

Yeah I have that portal open on my laptop quite often to see at least something.
